staging/src/k8s.io/component-base/logs/json/json.go

```diff
@@ -46,6 +47,48 @@ type zapLogger struct {
 // implement logr.Logger
 var _ logr.Logger = &zapLogger{}
 
+// getCallerFrame gets caller frame. The argument skip is the number of stack
+// frames to ascend, with 0 identifying the caller of getCallerFrame. The
+// boolean ok is false if it was not possible to recover the information.
+//
+// Note: This implementation is similar to runtime.Caller, but it returns the whole frame.
+func getCallerFrame(skip int) (frame runtime.Frame, ok bool) {
+	const skipOffset = 2 // skip getCallerFrame and Callers
+
+	pc := make([]uintptr, 1)
+	numFrames := runtime.Callers(skip+skipOffset, pc)
+	if numFrames < 1 {
+		return
+	}
+
+	frame, _ = runtime.CallersFrames(pc).Next()
+	return frame, frame.PC != 0
+}
+
+// try and set the Caller attribute of the entry,
+// if not possible, return silently.
+func attemptSettingCaller(skip int, entry *zapcore.Entry) {
+	// Get the frame of function that is at a depth of
+	// skip. For functions that are exposed such as Info
+	// and Error, this value is 2: depth of 1 frame for
+	// these logging functions and another depth 1 frame
+	// for the original caller. But for functions such as
+	// dPanic, skip will be 4, since this is used inside
+	// handleFields, which is used inside Write resulting
+	// in an additional depth of 2 frames.
```

Comment:
I think it's better arguments like `skip` in the function documentation and not comment.
```suggestion
```